### PR TITLE
DOC Add Exponentiated Gradient in reductions.rst

### PR DESCRIPTION
**Pull request by Hrittik20**
Issue: #1263

### CODE REVIEW

File: docs/user_guide/mitigation/reductions.rst [hunk 1]
```diff
@@ -557,6 +557,39 @@ Group :code:`"a"` has an average loss of :math:`0.05`, while group
...
+satisfy fairness constraints without needing access to sensitive features at deployment time.
+This algorithm creates a sequence of re-weighted datasets and retrains the 
+wrapped classifier on each of these datasets.
+To instantiate an ExponentiatedGradient model, you need to pass in a base estimator 
+and fairness constraints. The fairness constraints can be specified using an epsilon value,
+which represents the maximum allowed difference or ratio between the largest and smallest value.
+
+Here is an example of how to instantiate an ExponentiatedGradient model:
+
+.. doctest:: mitigation_reductions
```
Comment (MiroDudik) [inline docs/user_guide/mitigation/reductions.rst:573]: Instead of encapsulating inside `get_expgrad_models_per_epsilon`, let's just present this as a script. First, you could just call ` exp_grad_est.fit()`. Then report (or possibly plot) the (test or train) performance (in terms of accuracy and fairness) of the resulting ensemble, based on the predictions produced by `exp_grad_est.predict()`.

Then in the second part of the example, you can extract predictors `exp_grad_est.predictors_` and plot fairness vs accuracy of the individual models in the ensemble using [plot_model_comparison()](https://fairlearn.org/v0.10/api_reference/generated/fairlearn.metrics.plot_model_comparison.html)--perhaps also adding the ensemble predictions.

File: docs/user_guide/mitigation/reductions.rst [hunk 2]
```diff
@@ -557,6 +557,39 @@ Group :code:`"a"` has an average loss of :math:`0.05`, while group
     detected by :code:`gamma` is identical to the mean absolute error.
 
 
+Exponentiated Gradient
+----------------------
+
+The ExponentiatedGradient algorithm in Fairlearn is used to produce models that
```
Comment (romanlutz) [inline docs/user_guide/mitigation/reductions.rst:563]: Not: use rest syntax, like elsewhere in this file for classes.

**Comment by TamaraAtanasoska** [comment]
@hildeweerts I see you've merged the current `main` state into this PR, but I suppose it needs some syntax changes (and addressing @MiroDudik's comments) to make it merge properly. I am downloading it locally right now, but I can only make a separate PR with the changes since no pushing rights(I will reference this one so we can close it). I had some WIP that should expand this part of the documentation, but it isn't something I can work on within the next month so it might make sense to merge this as a start.

**Comment by TamaraAtanasoska** [comment]
@MiroDudik and @romanlutz, I just added a minimal example following the suggestions so we could merge this PR in [8d35cf0](https://github.com/fairlearn/fairlearn/pull/1353/commits/8d35cf003a1bde2bed9ba7b73de67a6b07cf3c94). Would that do? Is it correct? Here it is under just as a script so you could test it.

```
from fairlearn.datasets import fetch_adult
from fairlearn.metrics import plot_model_comparison, equal_opportunity_difference
from fairlearn.reductions import ExponentiatedGradient, EqualizedOdds
from sklearn.ensemble import  RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Fetch and preprocess the data
X, y = fetch_adult(return_X_y=True, as_frame=True)
A = X["sex"]

# Identify features
categorical_features = X.select_dtypes(include='category').columns.tolist()
numeric_features = X.select_dtypes(include='number').columns.tolist()

# Create preprocessor for X
preprocessor = ColumnTransformer(
    transformers=[
        ('num', 'passthrough', numeric_features),
        ('cat', OneHotEncoder(drop='first'), categorical_features)
    ])

# Transform y to numerical values
le = LabelEncoder()
y = le.fit_transform(y)

# Create a pipeline with the preprocessor and estimator
estimator = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(n_estimators=10, random_state=42))
])

# Split the data
X_train, X_test, y_train, y_test, A_train, A_test = train_test_split(X, y, A, test_size=0.2, random_state=42)

print("Training and evaluating base model...")
# Train and evaluate the base model
estimator.fit(X_train, y_train)
y_pred_base = estimator.predict(X_test)

# Create a list of ExponentiatedGradient models with different epsilons
epsilons = [0.001, 0.01, 0.05, 0.1, 0.2]
exp_grad_models = {}

print("Training and evaluating the ExponentiatedGradient models...")

for i, eps in enumerate(epsilons):
    exp_grad_est = ExponentiatedGradient(
        estimator=estimator,
        constraints=EqualizedOdds(difference_bound=eps),
        sample_weight_name="classifier__sample_weight",
    )
    exp_grad_est.fit(X_train, y_train, sensitive_features=A_train)
    exp_grad_models[f"ExpGrad (ε={eps})"] = exp_grad_est.predict(X_test)
    print(i+1)

# Add the base model predictions
exp_grad_models["Base Model"] = y_pred_base

print("Generating model comparison plot...")

# Plot the comparison 
plot_model_comparison(
    x_axis_metric=accuracy_score,
    y_axis_metric=equal_opportunity_difference,
    y_true=y_test,
    y_preds=exp_grad_models,
    sensitive_features=A_test,
    show_plot=True,
    point_labels =True,
)

```

File: docs/user_guide/mitigation/reductions.rst [hunk 3]
```diff
@@ -551,12 +551,88 @@ Group :code:`"a"` has an average loss of :math:`0.05`, while group
 
 .. note::
 
-    In the example above the :code:`BoundedGroupLoss` object does not use the
+    In the example above the :code::class:BoundedGroupLoss` object does not use the
```
Comment (romanlutz) [inline docs/user_guide/mitigation/reductions.rst:554]: The syntax looks off. Or does this work?

**Review by romanlutz — APPROVED** [review]
From my perspective this is great! @MiroDudik may have more thoughts.

If it takes too long to run you could always use fewer epsilon values, of course.

File: docs/user_guide/mitigation/reductions.rst (hunk 3 above)
Comment (TamaraAtanasoska) [reply docs/user_guide/mitigation/reductions.rst:554]: oh my, that is some weird editor hiccup that I didn't catch somehow by checking the diff. very likely a misplaced paste :S. it should just stay as it was for now :) thanks for catching it!

**Comment by TamaraAtanasoska** [comment]
> From my perspective this is great! @MiroDudik may have more thoughts.
> 
> If it takes too long to run you could always use fewer epsilon values, of course.

Just a comment that the version in the `reductions.rst` doesn't have any of the prints I wrote in my script before pasting it there :)

**Comment by TamaraAtanasoska** [comment]
@adrinjalali do PRs like this one matter for the release?

**Comment by adrinjalali** [comment]
@TamaraAtanasoska no, they're not a blocker.

**Comment by TamaraAtanasoska** [comment]
@adrinjalali can I improve the `scikit-learn` part of the example above? Better models or library concepts I could use?

**Comment by adrinjalali** [comment]
Modern examples should either be using `HistGradientBoosting*` for general purpose stuff, or probably `Ridge` or `SGD*`. `HistGradientBoosting*` are the best bet (and they're basically xgboost/lightgbm)

File: docs/user_guide/mitigation/reductions.rst [hunk 4]
```diff
@@ -557,6 +557,82 @@ Group :code:`"a"` has an average loss of :math:`0.05`, while group
...
+Exponentiated Gradient
+----------------------
+
+The :class:`ExponentiatedGradient` algorithm in Fairlearn is used to produce models that
+satisfy fairness constraints without needing access to sensitive features at deployment time.
+This algorithm creates a sequence of re-weighted datasets and retrains the
+wrapped classifier on each of these datasets.
+To instantiate an :class:`ExponentiatedGradient` model, we need to pass in a base estimator
+and fairness constraints. The fairness constraints can be specified using an epsilon value,
+which represents the maximum allowed difference or ratio between the largest and smallest value.
```
Comment (MiroDudik) [inline docs/user_guide/mitigation/reductions.rst:569]:
```suggestion
and fairness constraints. The fairness constraints are typically specified by providing
an upper bound on the difference (or the ratio) between the largest and the smallest
value of some statistic (like a false positive rate) across all groups. This bound is
often referred to as epsilon.
```

**Review by MiroDudik — COMMENTED** [review]
Just added a small rephrasing around the constraints. With that incorporated it's good to go as is.
